File: src/web/Controllers/PendingPositionsController.cs
using System;
using System.Threading.Tasks;
using core.fs.Stocks.PendingPositions;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using web.Utils;

namespace web.Controllers;

[ApiController]
[Authorize]
[Route("api/stocks/pendingpositions")]
public class PendingPositionsController(PendingStockPositionsHandler pendingStockPositionsHandler) : ControllerBase
{
    [HttpGet]
    public Task<ActionResult> PendingStockPositions() =>
        this.OkOrError(pendingStockPositionsHandler.Handle(new Query(User.Identifier())));
        
    [HttpGet("export")]
    public Task<ActionResult> ExportPendingStockPositions() =>
        this.GenerateExport(
            pendingStockPositionsHandler.Handle(new Export(User.Identifier())
            ));
        
    [HttpPost]
    public Task<ActionResult> CreatePendingStockPosition([FromBody]Create command) =>
        this.OkOrError(
            pendingStockPositionsHandler.HandleCreate(
                User.Identifier(), command
            )
        );

    [HttpPost("{id}/close")]
    public Task<ActionResult> ClosePendingStockPosition([FromBody] Close cmd) =>
        this.OkOrError(pendingStockPositionsHandler.Handle(cmd, User.Identifier()));
}

File: src/web.fs/Controllers/PendingPositionsController.fs
namespace web.Controllers

open System
open System.Threading.Tasks
open Microsoft.AspNetCore.Mvc
open Microsoft.AspNetCore.Authorization
open core.fs.Stocks.PendingPositions
open web.Utils

[<ApiController>]
[<Authorize>]
[<Route("api/stocks/pendingpositions")>]
type PendingPositionsController(pendingStockPositionsHandler: PendingStockPositionsHandler) =
    inherit ControllerBase()

    [<HttpGet>]
    member this.PendingStockPositions() =
        this.OkOrError(pendingStockPositionsHandler.Handle({Query.UserId = this.User.Identifier()}))

    [<HttpGet("export")>]
    member this.ExportPendingStockPositions() =
        this.GenerateExport(
            pendingStockPositionsHandler.Handle({Export.UserId = this.User.Identifier()})
        )

    [<HttpPost>]
    member this.CreatePendingStockPosition([<FromBody>] command: Create) =
        this.OkOrError(
            pendingStockPositionsHandler.HandleCreate (this.User.Identifier()) command
        )

    [<HttpPost("{id}/close")>]
    member this.ClosePendingStockPosition([<FromBody>] cmd: Close) =
        this.OkOrError(
            pendingStockPositionsHandler.Handle (cmd, this.User.Identifier())
        )
